### ISSUE TITLE
Random Data Provider

### ISSUE THREAD

**Issue by imacoindev** [EspeoBlockchain/gardener-server#22]
Gardener-server should deliver random data when requested.

**Comment by kss-espeo** [comment]
Due to blockchain's transparency and deterministic nature, it's not possible to provide a true random or even "good enough" pseudo-random number while working onchain. Still, there are business cases that require such number and an Oracle such as Gardener could provide them with minimal effort, as it already has a mechanism of safely providing offchain data.

**Comment by kss-espeo** [comment]
tl;dr:
We will implement it on SGX with a fallback mechanism of using 3rd party service in [random.org
](random.org)

@starspire and I decided that we will use sgx_read_rand function executed inside a SGX enclave to generate a random value. 

In theory, this is a pseudo-random value that could be reverse-engineered if a seed (which is most likely current cpu time, to be confirmed) was known to a malicious user. In practice however, we can't predict clock time at sgx_read_rand execution, as time elapsed between calling an enclave function and sgx_read_rand execution is non-deterministic. This, along with a fact that enclave code can't be tampered with as it is unmodifiable after it's been loaded practically ensures true-randomness. On top of that, we will be able to prove that it has been ran as intended using Intel Attestation Service.

This solution will require SGX on server machine. We will also implement a (configurable between sgx and fallback) fallback mechanism that doesn't require SGX. It will use open API of [random.org
](random.org) that produces true-random values. This solution requires user to trust 3rd party in random.org and it will also not be feasible (probably, this could be further researched) to prove that it was ran unpromised. But hey, this is just a fallback strategy so we don't care!
